<b> Fauzan Firdaus (23520011)<br> 
Nadya Aditama (23520039) </b>

# Flower Recognition using Decision Tree Learning, Random Forest, and XGBoost
Pada notebook ini, akan dilakukan eksperimen mengenai rekognisi citra bunga dengan melakukan 12 konfigurasi parameter dari algoritma machine learning Decision Tree Learning, Random Forest, dan XGBoost.




## Deskripsi Singkat Dataset
![CaptureFlower](https://user-images.githubusercontent.com/15353477/98227599-14368f80-1f8a-11eb-9af1-549adb75746f.PNG)
Dataset yang digunakan adalah dataset **Oxford 17 Flower**, terdiri dari 1**7 spesies**, dimana setiap spesies terdiri dari **80 citra**, sehingga total datasetnya adalah **1.360 citra**. Karakteristik data tersebut adalah sebagai berikut


*   Citra memiliki ukuran yang bervariasi, namun perbedaan ukurannya tidak jauh (kisaran 500-800 piksel).
*   Skala citra bervariasi (portrait, landscape, dan square).
*   Posture dan lightning dari citra bervariasi.
*   Dalam satu citra bisa terdiri dari lebih dari satu bunga.

## Import Package
Berikut ini adalah package yang dibutuhkan untuk eksperimen ini

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import cv2
import os
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from skimage.feature import hog
from skimage.feature import local_binary_pattern
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.tree import DecisionTreeClassifier
import xgboost as xgb
%matplotlib inline

## Load Image Data
Sebelum dilakukan load data, file citra dikategorikan dalam folder berdasarkan kelas. Setiap gambar disamakan ukurannya menjadi 300 x 300 piksel. Kemudian di-load dengan code berikut

In [2]:
DATADIR="17flowers"
CATEGORIES=["Bluebell","Buttercup","ColtsFoot",
            "Cowslip","Crocus","Daffodils",
            "Daisy","Dandelion","Fritillary",
            "Iris","LilyValley","Pansy",
            "Snowdrop","Sunflower","Tigerlily",
            "Tulip","Wildflower"]
training_data=[]
i = 1
for category in CATEGORIES:
    path = os.path.join(DATADIR,category)
    class_num=CATEGORIES.index(category)
    for img in os.listdir(path):
        try:
            img_array=cv2.imread(os.path.join(path,img))
            new_image=np.asarray(cv2.resize(img_array,(300,300)))
            training_data.append([new_image, class_num])
            print("gambar ke ",i)
            i=i+1
        except Exception as e:
            pass

gambar ke  1
gambar ke  2
gambar ke  3
gambar ke  4
gambar ke  5
gambar ke  6
gambar ke  7
gambar ke  8
gambar ke  9
gambar ke  10
gambar ke  11
gambar ke  12
gambar ke  13
gambar ke  14
gambar ke  15
gambar ke  16
gambar ke  17
gambar ke  18
gambar ke  19
gambar ke  20
gambar ke  21
gambar ke  22
gambar ke  23
gambar ke  24
gambar ke  25
gambar ke  26
gambar ke  27
gambar ke  28
gambar ke  29
gambar ke  30
gambar ke  31
gambar ke  32
gambar ke  33
gambar ke  34
gambar ke  35
gambar ke  36
gambar ke  37
gambar ke  38
gambar ke  39
gambar ke  40
gambar ke  41
gambar ke  42
gambar ke  43
gambar ke  44
gambar ke  45
gambar ke  46
gambar ke  47
gambar ke  48
gambar ke  49
gambar ke  50
gambar ke  51
gambar ke  52
gambar ke  53
gambar ke  54
gambar ke  55
gambar ke  56
gambar ke  57
gambar ke  58
gambar ke  59
gambar ke  60
gambar ke  61
gambar ke  62
gambar ke  63
gambar ke  64
gambar ke  65
gambar ke  66
gambar ke  67
gambar ke  68
gambar ke  69
gambar ke  70
gambar ke  71
gambar ke  72
g

gambar ke  555
gambar ke  556
gambar ke  557
gambar ke  558
gambar ke  559
gambar ke  560
gambar ke  561
gambar ke  562
gambar ke  563
gambar ke  564
gambar ke  565
gambar ke  566
gambar ke  567
gambar ke  568
gambar ke  569
gambar ke  570
gambar ke  571
gambar ke  572
gambar ke  573
gambar ke  574
gambar ke  575
gambar ke  576
gambar ke  577
gambar ke  578
gambar ke  579
gambar ke  580
gambar ke  581
gambar ke  582
gambar ke  583
gambar ke  584
gambar ke  585
gambar ke  586
gambar ke  587
gambar ke  588
gambar ke  589
gambar ke  590
gambar ke  591
gambar ke  592
gambar ke  593
gambar ke  594
gambar ke  595
gambar ke  596
gambar ke  597
gambar ke  598
gambar ke  599
gambar ke  600
gambar ke  601
gambar ke  602
gambar ke  603
gambar ke  604
gambar ke  605
gambar ke  606
gambar ke  607
gambar ke  608
gambar ke  609
gambar ke  610
gambar ke  611
gambar ke  612
gambar ke  613
gambar ke  614
gambar ke  615
gambar ke  616
gambar ke  617
gambar ke  618
gambar ke  619
gambar ke  620
gambar ke 

gambar ke  1099
gambar ke  1100
gambar ke  1101
gambar ke  1102
gambar ke  1103
gambar ke  1104
gambar ke  1105
gambar ke  1106
gambar ke  1107
gambar ke  1108
gambar ke  1109
gambar ke  1110
gambar ke  1111
gambar ke  1112
gambar ke  1113
gambar ke  1114
gambar ke  1115
gambar ke  1116
gambar ke  1117
gambar ke  1118
gambar ke  1119
gambar ke  1120
gambar ke  1121
gambar ke  1122
gambar ke  1123
gambar ke  1124
gambar ke  1125
gambar ke  1126
gambar ke  1127
gambar ke  1128
gambar ke  1129
gambar ke  1130
gambar ke  1131
gambar ke  1132
gambar ke  1133
gambar ke  1134
gambar ke  1135
gambar ke  1136
gambar ke  1137
gambar ke  1138
gambar ke  1139
gambar ke  1140
gambar ke  1141
gambar ke  1142
gambar ke  1143
gambar ke  1144
gambar ke  1145
gambar ke  1146
gambar ke  1147
gambar ke  1148
gambar ke  1149
gambar ke  1150
gambar ke  1151
gambar ke  1152
gambar ke  1153
gambar ke  1154
gambar ke  1155
gambar ke  1156
gambar ke  1157
gambar ke  1158
gambar ke  1159
gambar ke  1160
gambar k

## Split Train and Test Data
Dataset dibagi sebesar 80% data latih dan 20% data uji

In [3]:
X = []
y = []
for image,label in training_data:
    X.append(image)
    y.append(label)

In [4]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=42)

Jumlah data latih dan data uji adalah sebagai berikut

In [5]:
print("Jumlah data latih : ",np.shape(X_train)[0])
print("Jumlah data uji   : ",np.shape(X_test)[0]) 

Jumlah data latih :  1088
Jumlah data uji   :  272


## Preprocessing
Fitur yang akan dilihat dalam pengenalan citra bunga adalah fitur warna, bentuk, dan tekstur. Preprocessing yang dilakukan dalam citra tersebut adalah sebagai berikut
1.   Konversi ruang warna RGB ke Grayscale
2.   Konversi ruang warna RGB ke YCBcr



In [6]:
def preprocessing1(arr):
    arr_prep=[]
    for i in range(np.shape(arr)[0]):
        img=cv2.cvtColor(arr[i], cv2.COLOR_BGR2RGB)
        img=cv2.cvtColor(img, cv2.COLOR_RGB2GRAY)
        arr_prep.append(img)
    return arr_prep

In [7]:
def preprocessing2(arr):
    arr_prep=[]
    for i in range(np.shape(arr)[0]):
        image=cv2.cvtColor(arr[i], cv2.COLOR_BGR2GRAY)
        arr_prep.append(image)
    return arr_prep

In [8]:
def preprocessing3(arr):
    arr_prep=[]
    for i in range(np.shape(arr)[0]):
        image=cv2.cvtColor(arr[i], cv2.COLOR_BGR2YCrCb)
        arr_prep.append(image)
    return arr_prep

## Feature Extraction
Metode ekstraksi fitur yang digunakan dalam eksperimen ini adalah sebagai berikut
1.   HOG (Histogram of Oriented Gradient) sebagai deskriptor bentuk.
2.   LBP (Local Binary Pattern) sebagai deskriptor tesktur
3.   YCBcr (Color Histogram dari YCBcr) sebagai deskriptor warna



In [9]:
def FtrExtractHOG(img):
    ftr,_=hog(img, orientations=9, pixels_per_cell=(50, 50),
            cells_per_block=(2, 2), visualize=True, multichannel=False, block_norm= 'L2')
    return ftr

In [10]:
def FtrExtractLBP(img):
    radius = 1
    n_points = 8
    
    lbp = local_binary_pattern(img, n_points, radius, method='uniform')
    n_bins = int(lbp.max() + 1)
    hist, _ = np.histogram(lbp.ravel(), bins=n_bins, range=(0, n_bins))
    hist = hist.astype("float")                         
    hist /= (hist.sum() + (1e-7))
    return hist

In [11]:
def FtrExtractYcbcr(img):
    features = []

    for i in range(0,3):
        hist = cv2.calcHist([img],[i],None,[256],[0,256])
                        
    return np.array(hist).flatten()

## Proses Preprocessing dan Ekstraksi Fitur
Berikut ini adalah proses preprocessing dan ekstraksi fitur

In [12]:
def featureExtraction1(arr):
    arr_feature=[]
    for i in range(np.shape(arr)[0]):
        arr_feature.append(FtrExtractHOG(arr[i]))
    return arr_feature

def featureExtraction2(arr):
    arr_feature=[]
    for i in range(np.shape(arr)[0]):
        arr_feature.append(FtrExtractLBP(arr[i]))
    return arr_feature

def featureExtraction3(arr):
    arr_feature=[]
    for i in range(np.shape(arr)[0]):
        arr_feature.append(FtrExtractYcbcr(arr[i]))
    return arr_feature

Kemudian dilakukan preprocessing terhadap masing-masing jenis ekstraksi fitur

In [13]:
X_trainp_shape=preprocessing1(X_train)
X_testp_shape=preprocessing1(X_test)

X_trainp_texture=preprocessing2(X_train)
X_testp_texture=preprocessing2(X_test)

X_trainp_color=preprocessing3(X_train)
X_testp_color=preprocessing3(X_test)

Ekstraksi fitur bentuk, tekstur, dan warna

In [14]:
X_trainftr_shape=featureExtraction1(X_trainp_shape)
X_testftr_shape=featureExtraction1(X_testp_shape)

X_trainftr_texture=featureExtraction2(X_trainp_texture)
X_testftr_texture=featureExtraction2(X_testp_texture)

X_trainftr_color=featureExtraction3(X_trainp_color)
X_testftr_color=featureExtraction3(X_testp_color)

In [15]:
X_trainftr_combined = np.concatenate((X_trainftr_shape,X_trainftr_texture,X_trainftr_color),axis=1)
X_testftr_combined = np.concatenate((X_testftr_shape,X_testftr_texture,X_testftr_color),axis=1)

Ukuran dimensi fitur citra adalah sebagai berikut 

In [16]:
print("Dimensi fitur bentuk  : ",np.shape(X_trainftr_shape)[1])
print("Dimensi fitur tekstur : ",np.shape(X_trainftr_texture)[1])
print("Dimensi fitur warna   : ",np.shape(X_trainftr_color)[1])
print("Dimensi fitur gabungan: ",np.shape(X_trainftr_combined)[1])

Dimensi fitur bentuk  :  900
Dimensi fitur tekstur :  10
Dimensi fitur warna   :  256
Dimensi fitur gabungan:  1166


## Machine Learning Algorithm
Berikut ini adalah algoritma machine learning yang digunakan. Sebelum melakukan tunning parameter, akan dilihat hasil eksperimen per fitur

### Random Forest
Random Forest merupakan salah satu algoritma machine learning dengan konsep penggabungan pohon (tree) dengan melakukan training pada sample data yang dipilih.

#### Process Training
Proses training dilakukan per fitur, dan kemudian menggabungkan ketiga fitur tersebut

In [17]:
RF_clf1 = RandomForestClassifier(random_state = 42)
RF_clf1.fit(X_trainftr_shape, y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=42, verbose=0,
                       warm_start=False)

In [18]:
RF_clf2 = RandomForestClassifier(random_state = 42)
RF_clf2.fit(X_trainftr_texture, y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=42, verbose=0,
                       warm_start=False)

In [19]:
RF_clf3 = RandomForestClassifier(random_state = 42)
RF_clf3.fit(X_trainftr_color, y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=42, verbose=0,
                       warm_start=False)

In [20]:
RF_clf4 = RandomForestClassifier(random_state = 42)
RF_clf4.fit(X_trainftr_combined, y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=42, verbose=0,
                       warm_start=False)

#### Predict and Accuracy
Berikut ini adalah hasil prediksi dan akurasi yang dihasilkan dari algoritma **random forest** per deskriptor

In [21]:
y_pred_RF_shape =RF_clf1.predict(X_testftr_shape)
y_pred_RF_texture = RF_clf2.predict(X_testftr_texture)
y_pred_RF_color = RF_clf3.predict(X_testftr_color)
y_pred_RF_combined = RF_clf4.predict(X_testftr_combined)

In [22]:
print("Shape-RF Accuracy : ",accuracy_score(y_test, y_pred_RF_shape)*100,'%')
print("Texture-RF Accuracy : ",accuracy_score(y_test, y_pred_RF_texture)*100,'%')
print("Color-RF Accuracy : ",accuracy_score(y_test, y_pred_RF_color)*100,'%')
print("Combined-RF Accuracy : ",accuracy_score(y_test, y_pred_RF_combined)*100,'%')

Shape-RF Accuracy :  48.529411764705884 %
Texture-RF Accuracy :  17.647058823529413 %
Color-RF Accuracy :  45.588235294117645 %
Combined-RF Accuracy :  65.07352941176471 %


### Decision Tree Learning
Decision Tree merupakan salah satu algoritma machine learning dengan konsep mengubah data menjadi aturan-aturan keputusan **dalam bentuk tree**

#### Process Training
Proses training dilakukan per fitur, dan kemudian menggabungkan ketiga fitur tersebut

In [23]:
DTL_clf1 = DecisionTreeClassifier(random_state=42)
DTL_clf1.fit(X_trainftr_shape, y_train)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=42, splitter='best')

In [24]:
DTL_clf2 = DecisionTreeClassifier(random_state=42)
DTL_clf2.fit(X_trainftr_texture, y_train)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=42, splitter='best')

In [25]:
DTL_clf3 = DecisionTreeClassifier(random_state=42)
DTL_clf3.fit(X_trainftr_color, y_train)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=42, splitter='best')

In [26]:
DTL_clf4 = DecisionTreeClassifier(random_state=42)
DTL_clf4.fit(X_trainftr_combined, y_train)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=42, splitter='best')

#### Predict and Accuracy
Berikut ini adalah hasil prediksi dan akurasi yang dihasilkan dari algoritma **Decision Tree Learning** per deskriptor

In [27]:
y_pred_DTL_shape = DTL_clf1.predict(X_testftr_shape)
y_pred_DTL_texture = DTL_clf2.predict(X_testftr_texture)
y_pred_DTL_color = DTL_clf3.predict(X_testftr_color)
y_pred_DTL_combined = DTL_clf4.predict(X_testftr_combined)

In [28]:
print("Shape-DTL Accuracy : ",accuracy_score(y_test, y_pred_DTL_shape)*100,'%')
print("Texture-DTL Accuracy : ",accuracy_score(y_test, y_pred_DTL_texture)*100,'%')
print("Color-DTL Accuracy : ",accuracy_score(y_test, y_pred_DTL_color)*100,'%')
print("Combined-DTL Accuracy : ",accuracy_score(y_test, y_pred_DTL_combined)*100,'%')

Shape-DTL Accuracy :  25.36764705882353 %
Texture-DTL Accuracy :  19.11764705882353 %
Color-DTL Accuracy :  31.61764705882353 %
Combined-DTL Accuracy :  29.411764705882355 %


### XGBoost
XGBoost merupakan implementasi dari gradient boosting yang didesain untuk performansi dan kecepatan

#### Process Training
Proses training dilakukan per fitur, dan kemudian menggabungkan ketiga fitur tersebut

In [29]:
XG_clf1 = xgb.XGBClassifier()
XG_clf1.fit(np.asarray(X_trainftr_shape),y_train)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0,
              learning_rate=0.1, max_delta_step=0, max_depth=3,
              min_child_weight=1, missing=None, n_estimators=100, n_jobs=1,
              nthread=None, objective='multi:softprob', random_state=0,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
              silent=None, subsample=1, verbosity=1)

In [30]:
XG_clf2 = xgb.XGBClassifier()
XG_clf2.fit(np.asarray(X_trainftr_texture),y_train)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0,
              learning_rate=0.1, max_delta_step=0, max_depth=3,
              min_child_weight=1, missing=None, n_estimators=100, n_jobs=1,
              nthread=None, objective='multi:softprob', random_state=0,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
              silent=None, subsample=1, verbosity=1)

In [31]:
XG_clf3 = xgb.XGBClassifier()
XG_clf3.fit(np.asarray(X_trainftr_color),y_train)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0,
              learning_rate=0.1, max_delta_step=0, max_depth=3,
              min_child_weight=1, missing=None, n_estimators=100, n_jobs=1,
              nthread=None, objective='multi:softprob', random_state=0,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
              silent=None, subsample=1, verbosity=1)

In [32]:
XG_clf4 = xgb.XGBClassifier()
XG_clf4.fit(np.asarray(X_trainftr_combined),y_train)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0,
              learning_rate=0.1, max_delta_step=0, max_depth=3,
              min_child_weight=1, missing=None, n_estimators=100, n_jobs=1,
              nthread=None, objective='multi:softprob', random_state=0,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
              silent=None, subsample=1, verbosity=1)

#### Predict and Accuracy
Berikut ini adalah hasil prediksi dan akurasi yang dihasilkan dari algoritma **XGBoost** per deskriptor

In [33]:
y_pred_XG_shape = XG_clf1.predict(X_testftr_shape)
y_pred_XG_texture = XG_clf2.predict(X_testftr_texture)
y_pred_XG_color = XG_clf3.predict(X_testftr_color)
y_pred_XG_combined = XG_clf4.predict(X_testftr_combined)

In [34]:
print("Shape-XG Accuracy : ",accuracy_score(y_test, y_pred_XG_shape)*100,'%')
print("Texture-XG Accuracy : ",accuracy_score(y_test, y_pred_XG_texture)*100,'%')
print("Color-XG Accuracy : ",accuracy_score(y_test, y_pred_XG_color)*100,'%')
print("Combined-XG Accuracy : ",accuracy_score(y_test, y_pred_XG_combined)*100,'%')

Shape-XG Accuracy :  49.26470588235294 %
Texture-XG Accuracy :  19.852941176470587 %
Color-XG Accuracy :  41.17647058823529 %
Combined-XG Accuracy :  66.54411764705883 %


## Konfigurasi Parameter
Setelah itu, dilakukan pengujian eksperimen dengan tunning parameter untuk setiap algoritma sejumlah **12 konfigurasi**, dimana setiap algoritma dilakukan **4 konfigurasi**. Fitur yang dimasukkan adalah fitur yang telah digabungkan.

### Random Forest
Parameter yang ditunning dalam algoritma tersebut adalah sebagai berikut
1.   n_estimator (jumlah tree yang akan diestimasi dalam random forest)
2.   bootstrap (apakah sampel akan diambil dengan replacement atau tidak)

Adapun konfigurasi yang dilakukan adalah berdasarkan tabel berikut

| no | n_estimator | bootstrap |
| --- | --- | --- |
| 1 | 200 | True |
| 2 | 200 | False |
| 3 | 400 | True |
| 4 | 400 | False |

##### Random Forest Konfigurasi 1 (n_estimator=200, bootstrap=True)

In [35]:
RF1 = RandomForestClassifier(n_estimators=200, bootstrap=True, random_state = 42)
RF1.fit(X_trainftr_combined, y_train)
y_pred_RF1 = RF1.predict(X_testftr_combined)
print("RF-Conf1 Accuracy : ",accuracy_score(y_test, y_pred_RF1)*100,'%')

RF-Conf1 Accuracy :  65.80882352941177 %


##### Random Forest Konfigurasi 2 (n_estimator=200, bootstrap=False)

In [36]:
RF2 = RandomForestClassifier(n_estimators=200, bootstrap=False, random_state = 42)
RF2.fit(X_trainftr_combined, y_train)
y_pred_RF2 = RF2.predict(X_testftr_combined)
print("RF-Conf2 Accuracy : ",accuracy_score(y_test, y_pred_RF2)*100,'%')

RF-Conf2 Accuracy :  65.80882352941177 %


##### Random Forest Konfigurasi 3 (n_estimator=400, bootstrap=True)

In [37]:
RF3 = RandomForestClassifier(n_estimators=400, bootstrap=True, random_state = 42)
RF3.fit(X_trainftr_combined, y_train)
y_pred_RF3 = RF3.predict(X_testftr_combined)
print("RF-Conf3 Accuracy : ",accuracy_score(y_test, y_pred_RF3)*100,'%')

RF-Conf3 Accuracy :  66.17647058823529 %


##### Random Forest Konfigurasi 4 (n_estimator=400, bootstrap=False)

In [38]:
RF4 = RandomForestClassifier(n_estimators=400, bootstrap=False, random_state = 42)
RF4.fit(X_trainftr_combined, y_train)
y_pred_RF4 = RF4.predict(X_testftr_combined)
print("RF-Conf4 Accuracy : ",accuracy_score(y_test, y_pred_RF4)*100,'%')

RF-Conf4 Accuracy :  68.01470588235294 %


### Decision Tree Learning
Parameter yang ditunning dalam algoritma tersebut adalah sebagai berikut
1.   criterion (untuk pengukuran kualitas split)
2.   min_sample_split (jumlah minimal record yang diperbolehkan untuk di split lagi)

Adapun konfigurasi yang dilakukan adalah berdasarkan tabel berikut

| no | criterion | min_sample_split |
| --- | --- | --- |
| 1 | gini | 10 |
| 2 | gini | 20 |
| 3 | entropy | 10 |
| 4 | entropy | 20 |

##### Decision Tree Learning Konfigurasi 1 (criterion=gini, min_sample_split=10)

In [39]:
DTL1 = DecisionTreeClassifier(criterion="gini", min_samples_split=10, random_state=42)
DTL1.fit(X_trainftr_combined, y_train)
y_pred_DTL1 = DTL1.predict(X_testftr_combined)
print("DTL-Conf1 Accuracy : ",accuracy_score(y_test, y_pred_DTL1)*100,'%')

DTL-Conf1 Accuracy :  34.19117647058824 %


##### Decision Tree Learning Konfigurasi 2 (criterion=gini, min_sample_split=20)

In [40]:
DTL2 = DecisionTreeClassifier(criterion="gini", min_samples_split=20, random_state=42)
DTL2.fit(X_trainftr_combined, y_train)
y_pred_DTL2 = DTL2.predict(X_testftr_combined)
print("DTL-Conf2 Accuracy : ",accuracy_score(y_test, y_pred_DTL2)*100,'%')

DTL-Conf2 Accuracy :  32.35294117647059 %


##### Decision Tree Learning Konfigurasi 3 (criterion=entropy, min_sample_split=10)

In [41]:
DTL3 = DecisionTreeClassifier(criterion="entropy", min_samples_split=10, random_state=42)
DTL3.fit(X_trainftr_combined, y_train)
y_pred_DTL3 = DTL3.predict(X_testftr_combined)
print("DTL-Conf3 Accuracy : ",accuracy_score(y_test, y_pred_DTL3)*100,'%')

DTL-Conf3 Accuracy :  34.92647058823529 %


##### Decision Tree Learning Konfigurasi 4 (criterion=entropy, min_sample_split=20)

In [42]:
DTL4 = DecisionTreeClassifier(criterion="entropy", min_samples_split=20, random_state=42)
DTL4.fit(X_trainftr_combined, y_train)
y_pred_DTL4 = DTL4.predict(X_testftr_combined)
print("DTL-Conf4 Accuracy : ",accuracy_score(y_test, y_pred_DTL4)*100,'%')

DTL-Conf4 Accuracy :  37.86764705882353 %


### XGBoost
Parameter yang ditunning dalam algoritma tersebut adalah sebagai berikut
1.   subsample (persentase data yang diambil untuk membangun setiap tree)
2.   learning_rate (ukuran step untuk mencapai solusi yang optimal)

Adapun konfigurasi yang dilakukan adalah berdasarkan tabel berikut

| no | subsample | learning_rate |
| --- | --- | --- |
| 1 | 0.75 | 0.01 |
| 2 | 0.8 | 0.01 |
| 3 | 0.75 | 0.1 |
| 4 | 0.8 | 0.1 |

##### XGBoost Konfigurasi 1 (subsample=0.75, learning_rate=0.01)

In [43]:
XG1 = xgb.XGBClassifier(subsample=0.75, learning_rate=0.01)
XG1.fit(X_trainftr_combined, y_train)
y_pred_XG1 = XG1.predict(X_testftr_combined)
print("XG-Conf1 Accuracy : ",accuracy_score(y_test, y_pred_XG1)*100,'%')

XG-Conf1 Accuracy :  62.86764705882353 %


##### XGBoost Konfigurasi 2 (subsample=0.8, learning_rate=0.01)

In [44]:
XG2 = xgb.XGBClassifier(subsample=0.8, learning_rate=0.01)
XG2.fit(X_trainftr_combined, y_train)
y_pred_XG2 = XG2.predict(X_testftr_combined)
print("XG-Conf2 Accuracy : ",accuracy_score(y_test, y_pred_XG2)*100,'%')

XG-Conf2 Accuracy :  57.35294117647059 %


##### XGBoost Konfigurasi 3 (subsample=0.75, learning_rate=0.1)

In [45]:
XG3 = xgb.XGBClassifier(subsample=0.75, learning_rate=0.1)
XG3.fit(X_trainftr_combined, y_train)
y_pred_XG3 = XG3.predict(X_testftr_combined)
print("XG-Conf3 Accuracy : ",accuracy_score(y_test, y_pred_XG3)*100,'%')

XG-Conf3 Accuracy :  65.44117647058823 %


##### XGBoost Konfigurasi 4 (subsample=0.8, learning_rate=0.1)

In [46]:
XG4 = xgb.XGBClassifier(subsample=0.8, learning_rate=0.1)
XG4.fit(X_trainftr_combined, y_train)
y_pred_XG4 = XG4.predict(X_testftr_combined)
print("XG-Conf4 Accuracy : ",accuracy_score(y_test, y_pred_XG4)*100,'%')

XG-Conf4 Accuracy :  67.27941176470588 %


## Kesimpulan
*   Telah dilakukan eksperimen pengenalan citra bunga dengan menggunakan gabungan deskriptor fitur bentuk, tekstur, dan warna.
*   Telah dilakukan juga eksperimen dengan mempertimbangkan tiga algoritma, yaitu Random Forest, Decision Tree Learning, dan Random Forest, dan melakukan tunning parameter dengan total 12 konfigurasi untuk keseluruhan.
*   Untuk parameter terbaik dari setiap algoritma adalah sebagai berikut
    1. **Random Forest** dengan parameter **n_estimtator = 400** dan boostrap = false, menghasilkan akurasi **68.01%**
    2. **Decision Tree Learning** dengan parameter **criterion = "gini"** dan **min_sample_split = 20**, menghasilkan akurasi **37.8%**
    3. **XGBoost** dengan parameter **subsample = 0.8** dan **learning_rate = 0.1**, menghasilkan akurasi **67.27%**

